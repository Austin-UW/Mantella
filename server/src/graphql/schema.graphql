scalar Date

type Query {
  projects(ids: [String!]!): [Project]!
  projectById(id: String!): Project
  user(id: String!): User
}

type Tag {
  name: String!
  _id: String!
  color: String
}

type Task {
  name: String!
  points: Int!
  completed: Boolean!
  timeWorkedOn: Int!
  color: String # hex
  dueDate: Date # date string
  startDate: Date # date string
  assignedUsers: [String!]! # what if we had private tasks, that is a good idea, unique!
  tags: [String!]!
  subTasks: [SubTask!]!
  comments: [Comment!]!
  description: String
  recurrance: String # daily | weekly
  _id: String!
}

type User {
  _id: String!
  profileImg: String
  username: String!
  email: String!
  projects: [Project!]!
}

type Profile {
  _id: String!
  profileImg: String
  username: String!
  email: String!
  projects: [String!]!
}

type Project {
  ownerId: String
  name: String!
  _id: String!
  tags: [Tag!]
  columnIds: [String!]!
  columns: [Column!]!
  swimlanes: [Swimlane!]!
  users: [Profile!] # if null is a private project
  tasks: [Task!]!
  isPrivate: Boolean # defaults to false (public)
}

type Column {
  _id: String!
  name: String!
  isCompletedColumn: Boolean
  taskIds: [String!]!
  taskLimit: Int
}

type Swimlane {
  taskIds: [String!]!
  name: String!
  _id: String!
}

type SubTask {
  name: String!
  completed: Boolean!
  _id: String!
}

type Comment {
  description: String!
  dateAdded: Date! # date string
  lastEdited: Date # optional date string
  _id: String!
}

input TaskInput {
  name: String
  points: Int
  dueDate: Date # date string
  recurrance: String # daily | weekly | none
  color: String # hex
}

input ProjectInput {
  name: String
  columnIds: [String!]
  categories: [String!]
}

input ColumnInput {
  name: String
  isCompletedColumn: Boolean
  taskIds: [String!]
  taskLimit: Int
}

input SwimlaneInput {
  taskIds: [String!]
  name: String
}

input DragTaskIdList {
  id: String!
  newIds: [String!]!
}

type Auth {
  user: User!
}

type Void {
  message: String!
}

type DeleteReturn {
  id: String!
}

type TaskMerge {
  project: Project!
  task: Task
}

type ColumnMerge {
  project: Project!
  column: Column
}

type SwimlaneMerge {
  project: Project!
  swimlane: Swimlane
}

input SubtaskInfo {
  name: String
  completed: Boolean
}

type Mutation {
  createTask(
    projId: String!
    taskInfo: TaskInput!
    columnId: String!
  ): TaskMerge!
  editTask(projId: String!, task: TaskInput!, taskId: String!): TaskMerge!
  deleteTask(projId: String!, id: String!): TaskMerge!
  dragTask(
    columnIds: [DragTaskIdList!]!
    id: String!
    swimlaneIds: [DragTaskIdList!]!
    projectId: String!
  ): TaskMerge!

  # should these even be here?
  register(username: String!, password: String!, email: String!): Auth!
  login(email: String!, password: String!): Auth!
  loginWithCookie: Auth!
  logout: Void!

  createProject(name: String!): Project!
  editProject(id: String!, newProj: ProjectInput!): Project!
  deleteProject(id: String!): DeleteReturn! # sends deleted project id back
  createColumn(
    projId: String!
    name: String!
    isCompletedColumn: Boolean
    taskLimit: Int
  ): ColumnMerge!
  editColumn(
    colId: String!
    projectId: String!
    newCol: ColumnInput!
  ): ColumnMerge!
  deleteColumn(_id: String!, projectId: String!): ColumnMerge!

  joinProject(projectId: String!): Project!
  leaveProject(projectId: String!): DeleteReturn!

  createSwimlane(projId: String!, name: String!): SwimlaneMerge!
  editSwimlane(
    swimId: String!
    projId: String!
    newSwim: SwimlaneInput!
  ): SwimlaneMerge!
  deleteSwimlane(projId: String!, _id: String!): SwimlaneMerge!

  removeMemberFromProject(projectId: String!, userId: String!): Project!

  setComment(
    projId: String!
    taskId: String!
    commentId: String
    description: String!
  ): Task!
  setSubtask(
    projId: String!
    taskId: String!
    subtaskId: String
    info: SubtaskInfo
  ): Task!
}
