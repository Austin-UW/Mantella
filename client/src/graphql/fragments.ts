import gql from 'graphql-tag'

export type GQLExec = <T>(variables: T) => any

export const taskFields = gql`
  fragment taskFields on Task {
    points
    completed
    tags
    assignedUsers
    id: _id
    description
    dueDate
    startDate
    recurrance
    color
    subTasks {
      name
      completed
      id: _id
    }
    comments {
      id: _id
      description
      dateAdded
      lastEdited
    }
    timeWorkedOn
    name
  }
`

export const profileFields = gql`
  fragment ProfileFields on Profile {
    id: _id
    profileImg
    username
    email
    projects
  }
`

export const columnFields = gql`
  fragment columnFields on Column {
    id: _id
    name
    isCompletedColumn
    taskIds
    taskLimit
  }
`

export const projectFields = gql`
  ${taskFields}
  ${profileFields}
  ${columnFields}
  fragment projectFields on Project {
    columnIds
    ownerId
    columns {
      ...columnFields
    }
    swimlanes {
      taskIds
      name
      id: _id
    }
    users {
      ...ProfileFields
    }
    tasks {
      ...taskFields
    }
    id: _id
    name
    tags {
      name
      color
      id: _id
    }
  }
`

export const userFields = gql`
  ${projectFields}
  fragment userFields on User {
    id: _id
    profileImg
    username
    email
    projects {
      ...projectFields
    }
  }
`
