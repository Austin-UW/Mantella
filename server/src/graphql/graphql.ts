import { UserProps } from './../models/User'
import { ApolloServer } from 'apollo-server-express'

import fs from 'fs'
import path from 'path'
import { queries } from './queries/queries'
import { mutations } from './mutations/mutations'
import jsonwebtoken from 'jsonwebtoken'
import { UserModel } from '../models/User'
// Create a logger

const resolvers: any = {
  Query: queries,
  Mutation: mutations
}

const typeDefs: any = [
  fs.readFileSync(path.join(__dirname, 'schema.graphql'), 'utf8')
]

export const gqlServer = new ApolloServer({
  typeDefs,
  resolvers,
  engine: {
    apiKey: 'service:mantella:Ylb-gXcuSHCbWhBHn_Oa_A'
  },
  tracing: true,
  context: async ({ req, res }) => {
    try {
      const token: any = jsonwebtoken.decode(req.cookies['auth-token'])
      let user: UserProps | null | undefined

      if (token.id) {
        user = await UserModel.findOne({ _id: token.id })
      }

      return { req, res, user: user }
    } catch (err) {
      if (res) {
        res.clearCookie('auth-token')
      }

      return { req, res, user: null }
    }
  }
})
