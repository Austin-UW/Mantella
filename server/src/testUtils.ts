import { ProjectFieldsFragment } from './types'
import { MutationResult } from 'react-apollo'
import { ApolloServer } from 'apollo-server-express'
import { queries } from './graphql/queries/queries'
import { mutations } from './graphql/mutations/mutations'
import path from 'path'
import fs from 'fs'
import mongoose from 'mongoose'

// Load models since we will not be instantiating our express server.
mongoose.connect(
  process.env.DB_CONNECT ||
    'mongodb+srv://conrad:cokay101@kanbanbrawn-41gbz.mongodb.net/kanban?retryWrites=true',
  { useNewUrlParser: true }
)

mongoose.set('useFindAndModify', false)
mongoose.set('useCreateIndex', true)

const { createTestClient } = require('apollo-server-testing')

const resolvers: any = {
  Query: queries,
  Mutation: mutations
}

const typeDefs: any = [
  fs.readFileSync(path.join(__dirname, './graphql/schema.graphql'), 'utf8')
]

const gqlServer = new ApolloServer({
  typeDefs,
  resolvers,
  tracing: false,
  context: ({ req, res }) => {
    return {
      req,
      res,
      user: {
        projects: ['5d018bc58104190fe6e2ac2c'],
        _id: '5d018bc58104190fe6e2ac2d',
        password:
          '$2a$10$LL5wFKVW./hWctu8lypf/.geDiezQhQG6rYdbI7fIsVd0wfzYJwha',
        email: 'conradkay123@gmjail.com',
        username: 'kachow',
        profileImg:
          'https://mb.cision.com/Public/12278/2797280/879bd164c711a736_800x800ar.png',
        __v: 0
      }
    }
  }
})

export const test = createTestClient(gqlServer)

export const gqlReq = async <Q, V>(args: {
  query: any
  variables: V
  isMutation: boolean
  onError?: (err: any) => void
}): Promise<MutationResult<Q>> => {
  const result = args.isMutation
    ? await test.mutate({ mutation: args.query, variables: args.variables })
    : await test.query({ query: args.query, variables: args.variables })

  return result
}

export const projToSnapshot = (project: ProjectFieldsFragment) => {
  return {
    ...project,
    id: 'PROJECT_ID',
    columns: project.columns.map((column, i) => {
      return {
        ...column,
        id: `COL_ID ${i}`,
        taskIds: column.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    columnIds: project.columnOrder.map((columnId, i) => `COL_ID ${i}`),
    ownerId: 'OWNER_ID',
    swimlanes: project.swimlanes.map((swimlane, i) => {
      return {
        ...swimlane,
        id: `SWIM_ID ${i}`,
        taskIds: swimlane.taskIds.map((taskId, e) => `TASK_ID ${e}`)
      }
    }),
    tags: project.tags!.map((tag, i) => {
      return { ...tag, id: `TAG_ID ${i}` }
    }),
    tasks: project.tasks.map((task, i) => {
      return {
        ...task,
        id: `TASK_ID ${i}`,
        subTasks: task.subTasks.map((subTask, e) => {
          return { ...subTask, id: `SUBTASK_ID ${e}` }
        }),
        comments: task.comments.map((comment, e) => {
          return { ...comment, id: `COMMENT_ID ${e}` }
        })
      }
    }),
    users: project.users!.map((user, i) => {
      return { id: `USER_ID ${i}` }
    })
  }
}
