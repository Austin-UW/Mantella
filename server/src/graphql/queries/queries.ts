import { QueryResolvers } from '../../graphql/types'
import { ProjectModel } from '../../models/Project'

import { UserModel } from '../../models/User'
import mongoose from 'mongoose'
import { purifyProject, purifyProjects } from '../../utils'

export const queries: QueryResolvers = {
  projects: async (obj, args, context, info) => {
    const projects = await ProjectModel.find({
      _id: {
        $in: args.ids.map((id: string) => mongoose.Types.ObjectId(id))
      }
    })

    return (await purifyProjects(projects)) || []
  },
  projectById: async (obj, args, context, info) => {
    const proj = await ProjectModel.findById(args.id)

    if (proj) {
      return await purifyProject(proj)
    }

    return null
  },
  user: async (obj, args, context) => {
    const user = await UserModel.findById(args.id)

    if (user) {
      const userWithProjects = await user.populate('projects').execPopulate()

      const newProjects = await purifyProjects(userWithProjects.projects)

      const returning = {
        ...user.toObject(),
        projects: newProjects
      }

      return returning
    }
    return null
  }
}
