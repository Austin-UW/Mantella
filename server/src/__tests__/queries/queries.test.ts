import { GQL_GET_USER } from './../../../../client/src/graphql/queries'
import {
  ProjectQuery,
  ProjectQueryVariables,
  UserQuery,
  UserQueryVariables
} from './../../../../client/src/graphql/types'
import { gqlReq } from '../../testUtils'
import { GQL_GET_PROJECT } from '../../../../client/src/graphql/queries'

describe('GQL Queries', () => {
  test('Projects and projectById', async () => {
    const { data: projectData } = await gqlReq<
      ProjectQuery,
      ProjectQueryVariables
    >({
      isMutation: false,
      query: GQL_GET_PROJECT,
      variables: {
        id: '5d08fbce79d3d5773e47fd90'
      }
    })

    expect(projectData!.projectById!.id).toBe('5d08fbce79d3d5773e47fd90')
    expect(projectData!.projectById!).toMatchSnapshot()

    // ids: ['5d08fbce79d3d5773e47fd90', '5d08fbd9c53c0477859eeb56']
  })

  test('Gets user', async () => {
    const { data: userData } = await gqlReq<UserQuery, UserQueryVariables>({
      isMutation: false,
      query: GQL_GET_USER,
      variables: {
        id: '5d04064a9b99a535a2a28950'
      }
    })

    expect(userData!.user).toBeTruthy()

    expect(userData!.user!.id).toBe('5d04064a9b99a535a2a28950')
  })
})
