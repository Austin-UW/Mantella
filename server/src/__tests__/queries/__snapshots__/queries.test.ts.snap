// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GQL Queries Projects and projectById 1`] = `
Object {
  "columnIds": Array [
    "5d08fbce79d3d5773e47fd91",
  ],
  "columns": Array [
    Object {
      "id": "5d08fbce79d3d5773e47fd91",
      "isCompletedColumn": false,
      "name": "Todo",
      "taskIds": Array [],
      "taskLimit": 0,
    },
  ],
  "id": "5d08fbce79d3d5773e47fd90",
  "name": "PHOON_TIME",
  "ownerId": "5d04064a9b99a535a2a28950",
  "swimlanes": Array [],
  "tags": Array [],
  "tasks": Array [],
  "users": Array [],
}
`;
